### PR TITLE
Safer getattr

### LINKED ISSUE #142: `os.popen` Bypass
Hello! 

I just wanted to let you guys know that there is a work around for executing `os` module methods without the need of `__import__`. I'm aware that *RestrictedPython* is not a **"sandbox"** or a **"secure environment"**, but since there was **"Problematic code example"** in the docs, I just wanted to show you how this can be bypassed. I'm not really sure if this is a "bug" or not, but I just wanted to share :)

### Bypass
```python
>>> from RestrictedPython import compile_restricted
>>> from RestrictedPython import safe_builtins
>>>
>>> loc={}
>>> exec(compile_restricted("def exec_bypass(): return getattr(getattr(getattr(getattr(getattr([], '__class__'),'__base__'), '__subclasses__')().pop(78), '__init__'), '__globals__').pop('sys').modules.pop('os').popen('id').read()", '<inline>', 'exec'), safe_builtins, loc)
>>> loc['exec_bypass']()
'uid=0(root) gid=0(root) groups=0(root)\n'
```

Here, I used `getattr` since I couldn't access whatever that starts with an `"_"`. `78` is the index of `<class 'codecs.IncrementalEncoder'>` and this index can vary. If you are locally testing this out make sure to first find the index of the class `codecs.IncrementalEncoder`. I also had to use `pop` because `__getitem__` was not available.

This code kind of translates to the one below

```python
[].__class__.__base__.__subclasses__()[78].__init__.__globals__['sys'].modules['os'].popen("id").read()
```

Thanks!

### PR DESCRIPTION
**Pull request by loechel**
Proposal to fix #142

### CODE REVIEW

**Comment by loechel** [comment]
@icemac even if that might not be an attack vector for Zope and Plone could you please prioritize review, merge it yourself and cut a new release, please.

File: src/RestrictedPython/Guards.py [hunk 1]
```diff
@@ -266,10 +266,16 @@ def safer_getattr(object, name, getattr=getattr):
...
+    if name.startswith('_'):
+        raise AttributeError(
+            '"{name}" is an invalid attribute name because it '
+            'starts with "_"'.format(name=name)
+        )
+    return getattr(object, name, default)
 
 
 safe_builtins['_getattr_'] = safer_getattr
+safe_builtins['getattr'] = safer_getattr
```
Comment (icemac) [inline src/RestrictedPython/Guards.py]: I am not sure if this line is necessary. I will play with it and we will see.

File: tests/test_Guards.py [hunk 2]
```diff
@@ -241,3 +241,34 @@ def test_call_py2_builtins(c_exec):
...
+                        [], '__class__'
+                    ),
+                    '__base__'
+                ),
+                '__subclasses__'
+            )().pop(78),
+            '__init__'
+        ), '__globals__'
+    ).pop('sys').modules.pop('os').popen('id').read()
+"""
```
Comment (icemac) [inline tests/test_Guards.py]: This is the original code from the issue, but I think it is too complex.
`getattr([], '__class__')` should be enough as this is already forbidden.

File: src/RestrictedPython/Guards.py (hunk 1 above)
Comment (loechel) [reply src/RestrictedPython/Guards.py]: I have tested several ways, without this line it worked again in newer Python versions (3.6+)

**Comment by icemac** [comment]
@loechel Are you okay with my changes?

File: tests/test_Guards.py (hunk 2 above)
Comment (loechel) [reply tests/test_Guards.py]: For testing the new check, this is enough.

But the proposed bypass, also uses the pop method to gain access to certain modules, I would like to test that too, somewhere.

**Comment by icemac** [comment]
@loechel Released to PyPI: https://pypi.org/project/RestrictedPython/4.0b7/

**Pull request merged**